File: scripts/check-version-literals.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ALLOWED_FILES = [
  'src/lib/types/backend.ts',
  'src/lib/ai-contract.ts',
  'src/lib/ai-contract-version.ts',
  'src/lib/contract-registry.ts',
  'src/lib/services/analytics.service.ts',
];

const VERSION_LITERAL = /["']([1-6]\.0)["']/g;
const VERSION_IDENTIFIER =
  /AI_ANALYTICS_(?:V\d+_|DYNAMIC_|SUPPORTED_)?CONTRACT_VERSIONS?/;
const COMPARISON = /===|!==|==|!=|\.includes\s*\(/;

function isAllowedFile(filePath) {
  if (filePath.includes('__tests__')) return true;
  return ALLOWED_FILES.some((allowed) => filePath.endsWith(allowed));
}

export function findVersionBranching(source, filePath = '<source>') {
  const errors = [];
  const lines = source.split('\n');

  lines.forEach((line, index) => {
    VERSION_LITERAL.lastIndex = 0;
    let match;
    while ((match = VERSION_LITERAL.exec(line)) !== null) {
      const isJsonRpc = /jsonrpc/i.test(line);
      const isContractContext = /contract.?version/i.test(line);
      if (
        !isJsonRpc &&
        (isContractContext || Number(match[1]) >= 3 || COMPARISON.test(line))
      ) {
        errors.push(
          `${filePath}:${index + 1}: Found hardcoded contract version literal: ${match[0]}`,
        );
      }
    }

    if (VERSION_IDENTIFIER.test(line) && COMPARISON.test(line)) {
      errors.push(
        `${filePath}:${index + 1}: Found contract-version branching through a named constant`,
      );
    }
  });

  // Catch multiline membership checks such as
  // [AI_ANALYTICS_V4_CONTRACT_VERSION, ...].includes(version).
  const multilineBranch = new RegExp(
    `${VERSION_IDENTIFIER.source}[\\s\\S]{0,300}?\\.includes\\s*\\(`,
    'g',
  );
  for (const match of source.matchAll(multilineBranch)) {
    const line = source.slice(0, match.index).split('\n').length;
    const message = `${filePath}:${line}: Found contract-version branching through a named constant`;
    if (!errors.includes(message)) errors.push(message);
  }

  const multilineComparison = new RegExp(
    `(?:${VERSION_IDENTIFIER.source}[\\s\\S]{0,120}?(?:===|!==|==|!=)|` +
      `(?:===|!==|==|!=)[\\s\\S]{0,120}?${VERSION_IDENTIFIER.source})`,
    'g',
  );
  for (const match of source.matchAll(multilineComparison)) {
    const line = source.slice(0, match.index).split('\n').length;
    const message = `${filePath}:${line}: Found contract-version branching through a named constant`;
    if (!errors.includes(message)) errors.push(message);
  }

  return errors;
}

function scanDir(dir, errors) {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      scanDir(fullPath, errors);
    } else if (
      (fullPath.endsWith('.ts') || fullPath.endsWith('.tsx')) &&
      !isAllowedFile(fullPath)
    ) {
      errors.push(
        ...findVersionBranching(fs.readFileSync(fullPath, 'utf-8'), fullPath),
      );
    }
  }
}

function main() {
  const errors = [];
  scanDir('src', errors);
  if (errors.length > 0) {
    console.error('Architecture fitness check failed:');
    errors.forEach((error) => console.error(error));
    process.exit(1);
  }
  console.log('Architecture fitness check passed.');
}

if (process.argv[1] === fileURLToPath(import.meta.url)) main();

File: ai-analytics-service/scripts/check_version_literals.py
import ast
import os
import re
import sys


ALLOWED_FILES = (
    "src/contracts.py",
    "src/schemas/contract_registry.py",
    "src/schemas/mcp_types.py",
)
VERSION_IDENTIFIER = re.compile(
    r"^AI_ANALYTICS_(?:V\d+_|DYNAMIC_|SUPPORTED_)?CONTRACT_VERSIONS?$"
)


def _is_allowed_file(filepath):
    normalized = filepath.replace(os.sep, "/")
    return "tests/" in normalized or any(
        normalized.endswith(allowed) for allowed in ALLOWED_FILES
    )


def _contains_version_identifier(node):
    return any(
        isinstance(child, ast.Name)
        and VERSION_IDENTIFIER.match(child.id)
        for child in ast.walk(node)
    )


def check_python_source(source, filepath="<source>"):
    tree = ast.parse(source, filename=filepath)
    errors = []

    class Visitor(ast.NodeVisitor):
        def visit_Constant(self, node):
            if (
                isinstance(node.value, str)
                and re.fullmatch(r"[1-6]\.0", node.value)
                and int(node.value.split(".")[0]) >= 3
            ):
                errors.append(
                    f"{filepath}:{node.lineno}: literal '{node.value}' "
                    "found outside a contract boundary"
                )
            self.generic_visit(node)

        def visit_Compare(self, node):
            if _contains_version_identifier(node):
                errors.append(
                    f"{filepath}:{node.lineno}: contract-version branching "
                    "through a named constant"
                )
            for child in ast.walk(node):
                if (
                    isinstance(child, ast.Constant)
                    and isinstance(child.value, str)
                    and re.fullmatch(r"(?:1|2)\.0", child.value)
                ):
                    errors.append(
                        f"{filepath}:{node.lineno}: legacy contract-version "
                        f"branching on '{child.value}'"
                    )
            self.generic_visit(node)

    Visitor().visit(tree)
    return list(dict.fromkeys(errors))


def check_python_file(filepath):
    if _is_allowed_file(filepath):
        return []
    with open(filepath, "r", encoding="utf-8") as source_file:
        return check_python_source(source_file.read(), filepath)


def main():
    errors = []
    for root, _, files in os.walk("ai-analytics-service/src"):
        for filename in files:
            if filename.endswith(".py"):
                errors.extend(check_python_file(os.path.join(root, filename)))

    for error in errors:
        print(error)
    if errors:
        sys.exit(1)


if __name__ == "__main__":
    main()
